### PR TITLE
Add attributes benchmarking and a saveLayer benchmark to DisplayList benchmarks

### PR DESCRIPTION
**Pull request by gw280**
This adds a new saveLayer benchmark, and also adds the necessary changes to the harness to support benchmarking with different attributes.

The attributes are reported via the `benchmark` library's "counters" API - we report 1 if an attribute is enabled, and 0 if it isn't. Only attributes supported by the specific rasterop being benchmarked are reported.

For now, the attributes being benchmarked are:

* Hairline vs non-hairline stroke widths
* Stroked vs filled styles
* Anti-aliased vs non-anti-aliased

It should be relatively easy to add new attributes in the future.

### CODE REVIEW

File: display_list/display_list_benchmarks.cc [hunk 1]
```diff
@@ -732,10 +844,16 @@ sk_sp<SkImage> ImageFromBitmapWithNewID(const SkBitmap& bitmap) {
...
+                  unsigned attributes,
                   const SkSamplingOptions& options,
                   bool upload_bitmap) {
   auto canvas_provider = CreateCanvasProvider(backend_type);
   DisplayListBuilder builder;
+  builder.setAttributesFromPaint(GetPaintForRun(attributes),
+                                 DisplayListOpFlags::kDrawImageWithPaintFlags);
+  AnnotateAttributes(attributes, state,
+                     DisplayListOpFlags::kDrawImageWithPaintFlags);
+
```
[the remark this answers is not in the document]
Comment (flar) [reply display_list/display_list_benchmarks.cc:856]: It looks like you benchmark these with the empty attributes so that should cover it.

File: display_list/display_list_benchmarks.cc [hunk 2]
```diff
@@ -32,6 +33,45 @@ std::unique_ptr<CanvasProvider> CreateCanvasProvider(BackendType backend_type) {
   return nullptr;
 }
 
+SkPaint GetPaintForRun(unsigned attributes) {
+  SkPaint paint;
+
+  if (attributes & kStrokedStyle_Flag && attributes & kFilledStyle_Flag) {
+    paint.setStyle(SkPaint::kStrokeAndFill_Style);
```
Comment (flar) [inline display_list/display_list_benchmarks.cc:40]: Flutter doesn't have strokeAndFill, but it can't hurt to test it anyway as DL will accept that style.

**Review by flar — CHANGES_REQUESTED** [review]
Looks good except for drawColor in the saveLayers. That's a problematic corner case that I'm not sure we ever encounter in practice.

File: display_list/display_list_benchmarks.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (gw280) [reply display_list/display_list_benchmarks.cc:856]: Yes, everything should be benchmarked with empty attributes as well as any specialised ones

File: display_list/display_list_benchmarks.cc [hunk 3]
```diff
@@ -1060,5 +1203,50 @@ void BM_DrawShadow(benchmark::State& state,
...
+  auto canvas = canvas_provider->GetSurface()->getCanvas();
+
+  size_t save_layer_calls = state.range(0);
+
+  for (size_t i = 0; i < save_layer_calls; i++) {
+    for (size_t j = 0; j < save_depth; j++) {
+      builder.saveLayer(nullptr, false);
+      builder.drawColor(SK_ColorRED, SkBlendMode::kSrc);
+    }
+    for (size_t j = 0; j < save_depth; j++) {
```
[the remark this answers is not in the document]
Comment (gw280) [reply display_list/display_list_benchmarks.cc:1241]: Changed it to two overlapping rectangles for every saveLayer.

**Review by flar — APPROVED** [review]
LGTM
